### DISCUSSION TITLE
dynamic middleware

### DISCUSSION THREAD

**Discussion by MrOxMasTer** [next-safe-action/next-safe-action#261] [Ideas]
Just as a suggestion (use them as decorators). You can probably implement such middleware yourself, but I think there will be such middleware useful in the library itself (if my implementation turns out not to be very good)

Example of middleware for authorization:
```ts
const authMiddleware = (permissions) => createMiddleware().define(async ({ next }) => {

  const session = getSession()
  const permissionsInRole = getPermissionsInRole(session.role) // cached

  for(;;) {
    ...

   if (false) { // There is no necessary permission
         throw new Error('No permit') 
   }
  }
  return next()
})
```

other file:
```ts
const permissions = ['user:delete']

const deleteUser = actionClient().use(authMiddleware(permissions)).stateAction(...)

```

**Comment by TheEdoRan** [comment]
next-safe-action is not tied to any library, and the `createMiddleware` utility is very powerful, you can reuse middleware functions and even publish packages if you want. I could release a middleware-next-auth package, just like the react-hook-form adapter, but it's very simple to set up, so I don't think it makes much sense.
